### PR TITLE
dtype for encoders configuration

### PR DESCRIPTION
**Pull request by logogin**
#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [ ] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [ ] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [ ] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [ ] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/encoder.py [hunk 1]
```diff
@@ -51,16 +52,19 @@ class RecurrentEncoderConfig(config.Config):
     :param rnn_config: RNN configuration.
     :param conv_config: Optional configuration for convolutional embedding.
     :param reverse_input: Reverse embedding sequence before feeding into RNN.
+    :param dtype: data type.
```
Comment (fhieber) [inline sockeye/encoder.py]: Nit pick: Capital first letter on param docstrings

File: sockeye/encoder.py [hunk 2]
```diff
@@ -70,20 +74,23 @@ class ConvolutionalEncoderConfig(config.Config):
     :param cnn_config: CNN configuration.
     :param num_layers: The number of convolutional layers on top of the embeddings.
     :param positional_embedding_type: The type of positional embedding.
+    :param dtype: data type.
```
Comment (fhieber) [inline sockeye/encoder.py]: same here

File: sockeye/encoder.py [hunk 3]
```diff
@@ -164,26 +174,32 @@ def get_transformer_encoder(config: transformer.TransformerConfig, prefix: str)
...
+    encoder_seq.append(TransformerEncoder, config=config, prefix=prefix + C.TRANSFORMER_ENCODER_PREFIX)
 
-    return EncoderSequence(encoders)
+    return encoder_seq
 
 
 class Encoder(ABC):
     """
     Generic encoder interface.
+    :param dtype: data type.
```
Comment (fhieber) [inline sockeye/encoder.py]: missing newline

File: sockeye/transformer.py [hunk 4]
```diff
@@ -19,6 +19,7 @@
 from . import config
 from . import constants as C
 from . import layers
+from . import encoder
```
Comment (fhieber) [inline sockeye/transformer.py]: could this be a circular import thats causing you problems with Python 3.4?

File: sockeye/encoder.py [hunk 5]
```diff
@@ -95,64 +102,67 @@ def get_recurrent_encoder(config: RecurrentEncoderConfig, prefix: str) -> 'Encod
...
-                                                          scale_down_positions=False,
-                                                          prefix="%s%sadd_positional_encodings" %
-                                                                 (prefix, C.CHAR_SEQ_ENCODER_PREFIX)))
-        encoders.append(ConvertLayout(C.TIME_MAJOR, num_hidden=encoders[-1].get_num_hidden()))
+            encoder_seq.append(AddSinCosPositionalEmbeddings,
+                               num_embed=config.conv_config.num_embed,
+                               scale_up_input=False,
+                               scale_down_positions=False,
+                               prefix="%s%sadd_positional_encodings" %
+                                                                 (prefix, C.CHAR_SEQ_ENCODER_PREFIX))
```
Comment (fhieber) [inline sockeye/encoder.py]: [nit] indentation seems strange here

File: sockeye/encoder.py [hunk 6]
```diff
@@ -636,6 +684,28 @@ def get_max_seq_len(self) -> Optional[int]:
                            for encoder in self.encoders if encoder.get_max_seq_len() is not None), default=None)
         return max_seq_len
 
+    def append(self, cls, infer_hidden: bool = False, **kwargs) -> Encoder:
+        """
+        Extends sequence with new Encoder. 'dtype' gets gets passed into Encoder instance if not present in parameters
+        and supported by specific Encoder type.
+
+        :param cls: Encoder type
```
Comment (fhieber) [inline sockeye/encoder.py]: [nit] missing period.

**Comment by logogin** [comment]
any additional comments?

File: sockeye/encoder.py [hunk 7]
```diff
@@ -254,9 +272,11 @@ def get_num_hidden(self) -> int:
 class ReverseSequence(Encoder):
     """
     Reverses the input sequence. Requires time-major layout.
+    :param dtype: Data type.
```
Comment (fhieber) [inline sockeye/encoder.py]: missing newline

File: sockeye/encoder.py [hunk 8]
```diff
@@ -547,9 +576,11 @@ def get_max_seq_len(self) -> Optional[int]:
 class NoOpPositionalEmbeddings(PositionalEncoder):
     """
     Simple NoOp pos embedding. It does not modify the data, but avoids lots of if statements.
+    :param dtype: Data type.
```
Comment (fhieber) [inline sockeye/encoder.py]: missing newline

File: sockeye/encoder.py [hunk 9]
```diff
@@ -636,6 +684,28 @@ def get_max_seq_len(self) -> Optional[int]:
                            for encoder in self.encoders if encoder.get_max_seq_len() is not None), default=None)
         return max_seq_len
 
+    def append(self, cls, infer_hidden: bool = False, **kwargs) -> Encoder:
+        """
+        Extends sequence with new Encoder. 'dtype' gets gets passed into Encoder instance if not present in parameters
```
Comment (fhieber) [inline sockeye/encoder.py]: double 'gets'

File: sockeye/encoder.py [hunk 10]
```diff
@@ -636,6 +684,28 @@ def get_max_seq_len(self) -> Optional[int]:
...
+        and supported by specific Encoder type.
+
+        :param cls: Encoder type.
+        :param infer_hidden: if number of hidden should be inferred from previous encoder.
```
Comment (fhieber) [inline sockeye/encoder.py]: We capitalize the first first in param description.

File: sockeye/rnn.py [hunk 11]
```diff
@@ -36,6 +36,7 @@ class RNNConfig(Config):
     :param first_residual_layer: First layer with a residual connection (1-based indexes).
            Default is to start at the second layer.
     :param forget_bias: Initial value of forget biases.
+    :param dtype: data type.
```
Comment (fhieber) [inline sockeye/rnn.py]: capitalization

File: test/unit/test_encoder.py [hunk 12]
```diff
@@ -24,6 +25,96 @@
 _DATA_LENGTH_ND = mx.nd.array([1, 2, 3, 4, 5, 6, 7, 8])
 
 
+def test_get_recurrent_encoder():
```
Comment (fhieber) [inline test/unit/test_encoder.py]: Thanks for adding these tests! Could you add 1 test where we don't supply a conv_config (which is mostly the default)?

**Pull request merged**